File: ui/src/api/client-pages.js
/** @import { PageSite, PageSiteUpdate } from './types.js' */
import { SystemControllerClient } from './core.js'

/**
 * Create a new page site for static content hosting.
 *
 * @param {string} name - Unique name for the page site (required).
 * @param {string} repoURL - Git repository URL containing static content (required for source_type "git").
 * @param {string} branch - Git branch to serve. Defaults to "main" on the server if empty.
 * @param {string} [domain] - Custom domain for the page. Defaults to the page name if empty.
 * @param {string} [sourceType] - Content source: "archive" (default), "container_image", or "git".
 * @param {string} [image] - Container image reference (required for source_type "container_image").
 * @param {string} [imageDirectory] - Directory within the container image to extract (required for source_type "container_image").
 * @returns {Promise<PageSite>}
 *
 * Calls POST /pages/create on the Control Plane Service.
 */
SystemControllerClient.prototype.createPage = async function (name, repoURL, branch, domain, sourceType, image, imageDirectory) {
  return this.postJSON('/pages/create', {
    name,
    repo_url: repoURL || '',
    branch: branch || '',
    domain: domain || '',
    source_type: sourceType || 'archive',
    image: image || '',
    image_directory: imageDirectory || '',
  })
}

/**
 * Update fields on an existing page site. Only provided fields are changed.
 *
 * @param {string} name - Page site name to update (required).
 * @param {PageSiteUpdate} fields - Partial update: repo_url, branch, domain, source_type,
 *   image, image_directory, status. Only non-undefined fields are applied.
 * @returns {Promise<PageSite>}
 *
 * Calls POST /pages/update on the Control Plane Service.
 */
SystemControllerClient.prototype.updatePage = async function (name, fields) {
  return this.postJSON('/pages/update', { name, fields })
}

/**
 * Delete a page site and its cloned repository.
 *
 * @param {string} name - Page site name to remove (required).
 * @returns {Promise<void>}
 *
 * Calls POST /pages/remove on the Control Plane Service.
 */
SystemControllerClient.prototype.removePage = async function (name) {
  await this.post('/pages/remove', { name })
}

/**
 * List page sites with pagination, sorting, and search.
 *
 * @param {string} [sortBy] - Field to sort by (e.g. "name", "domain", "status", "source_type"). Default: server default (name).
 * @param {string} [sortOrder] - Sort direction: "asc" (default) or "desc".
 * @param {number} [limit] - Maximum entries per page. Default: server default (20).
 * @param {number} [offset] - Number of entries to skip for pagination. Default: 0.
 * @param {string} [search] - Case-insensitive search filter across all fields.
 * @returns {Promise<{entries: PageSite[], has_more: boolean, total_pages: number, total_count: number}>}
 *
 * Calls GET /pages on the Control Plane Service.
 */
SystemControllerClient.prototype.listPages = async function (sortBy, sortOrder, limit, offset, search) {
  const params = new URLSearchParams()
  if (sortBy) params.set('sort_by', sortBy)
  if (sortOrder) params.set('sort_order', sortOrder)
  if (limit) params.set('limit', String(limit))
  if (offset) params.set('offset', String(offset))
  if (search) params.set('search', search)
  const qs = params.toString()
  return this.getJSON(`/pages${qs ? `?${qs}` : ''}`)
}

/**
 * Rebuild a page by pulling the latest content from its git repository
 * (git pages), re-extracting from the container image (container_image pages),
 * or returning an error for archive pages (use uploadPageArchive instead).
 *
 * @param {string} name - Page site name to rebuild (required).
 * @returns {Promise<PageSite>}
 *
 * Calls POST /pages/rebuild on the Control Plane Service.
 */
SystemControllerClient.prototype.rebuildPage = async function (name) {
  return this.postJSON('/pages/rebuild', { name })
}

/**
 * Upload and extract an archive for an archive-type page. The page transitions
 * to "active" status on success or "error" on failure. Only valid for pages
 * with source_type "archive".
 *
 * @param {string} name - Page name to upload content for (required).
 * @param {File} file - Archive file. Supported formats: .tar, .tar.gz/.tgz,
 *   .tar.bz2/.tbz2, .tar.xz/.txz.
 * @returns {Promise<PageSite>}
 *
 * Calls POST /pages/upload on the Control Plane Service.
 */
SystemControllerClient.prototype.uploadPageArchive = async function (name, file) {
  const form = new FormData()
  form.append('name', name)
  form.append('archive', file)

  const resp = await fetch(`${this.baseURL}/pages/upload`, {
    method: 'POST',
    headers: { 'Authorization': `Bearer ${this.token}` },
    body: form,
  })

  if (!resp.ok) {
    const text = await resp.text()
    let detail = text
    try {
      const parsed = JSON.parse(text)
      detail = parsed.detail || parsed.message || text
    } catch { /* use raw text */ }
    const err = new Error(detail)
    err.detail = detail
    throw err
  }

  return resp.json()
}

File: src/svc/systemcontroller/client_pages.go
package systemcontroller

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"

	"gitea.com/town-os/town-os/src/account"
)

// CreatePage creates a new static site page. The repository is cloned
// asynchronously for git source types; container images are extracted
// asynchronously; archive pages remain pending until an archive is uploaded
// via UploadPageArchive.
//
// Parameters:
//   - name: unique identifier for the page (primary key).
//   - repoURL: git repository URL to clone (required for source_type "git").
//   - branch: branch to check out (defaults to "main" if empty).
//   - domain: domain or subdomain to serve the page on (defaults to name if empty).
//   - sourceType: content source — "archive" (default), "container_image", or "git".
//   - image: container image reference (required for source_type "container_image").
//   - imageDirectory: directory within the container image to extract (required for source_type "container_image").
//   - network: the network to publish the page on ("" = the default/home network).
//
// Calls POST /pages/create on the Control Plane Service.
func (c *SystemdClient) CreatePage(ctx context.Context, name, repoURL, branch, domain, sourceType, image, imageDirectory, network string) (_ *account.PageSite, err error) {
	pr, pw := io.Pipe()
	go pipeEncode(pw, CreatePageRequest{
		Name:           name,
		RepoURL:        repoURL,
		Branch:         branch,
		Domain:         domain,
		SourceType:     sourceType,
		Image:          image,
		ImageDirectory: imageDirectory,
		Network:        network,
	})

	resp, err := c.postJSON(ctx, "pages/create", pr)
	if err != nil {
		return nil, fmt.Errorf("%w: CreatePage: %w", ErrHTTPRequest, err)
	}
	defer func() {
		err = errors.Join(err, resp.Body.Close())
	}()

	if resp.StatusCode != http.StatusOK {
		return nil, readProblemDetail(resp, "POST", "pages/create")
	}

	var page account.PageSite
	return &page, json.NewDecoder(resp.Body).Decode(&page)
}

// UpdatePage performs a partial update of a page's fields. Only provided
// (non-nil) fields in the update struct are changed.
//
// Parameters:
//   - name: the page name to update.
//   - fields: partial update with optional RepoURL, Branch, Domain, SourceType,
//     Image, and ImageDirectory fields.
//
// Calls POST /pages/update on the Control Plane Service.
func (c *SystemdClient) UpdatePage(ctx context.Context, name string, fields account.PageSiteUpdate) (_ *account.PageSite, err error) {
	pr, pw := io.Pipe()
	go pipeEncode(pw, UpdatePageRequest{Name: name, Fields: fields})

	resp, err := c.postJSON(ctx, "pages/update", pr)
	if err != nil {
		return nil, fmt.Errorf("%w: UpdatePage: %w", ErrHTTPRequest, err)
	}
	defer func() {
		err = errors.Join(err, resp.Body.Close())
	}()

	if resp.StatusCode != http.StatusOK {
		return nil, readProblemDetail(resp, "POST", "pages/update")
	}

	var page account.PageSite
	return &page, json.NewDecoder(resp.Body).Decode(&page)
}

// RemovePage deletes a page and its cloned repository directory.
//
// Parameters:
//   - name: the page name to remove.
//
// Calls POST /pages/remove on the Control Plane Service.
func (c *SystemdClient) RemovePage(ctx context.Context, name string) error {
	pr, pw := io.Pipe()
	go pipeEncode(pw, PageNameRequest{Name: name})

	return c.postClient(ctx, "pages/remove", pr)
}

// ListPages returns a paginated list of all page sites with sorting, search,
// and pagination support via params.
//
// Parameters:
//   - params: sorting, pagination, and search options.
//
// Calls GET /pages on the Control Plane Service.
func (c *SystemdClient) ListPages(ctx context.Context, params ListParams) (_ *PageResult[account.PageSite], err error) {
	resp, err := c.getClient(ctx, "pages"+params.QueryString())
	if err != nil {
		return nil, fmt.Errorf("%w: ListPages: %w", ErrHTTPRequest, err)
	}
	defer func() {
		err = errors.Join(err, resp.Body.Close())
	}()

	if resp.StatusCode != http.StatusOK {
		return nil, readProblemDetail(resp, "GET", "pages")
	}

	var page PageResult[account.PageSite]
	return &page, json.NewDecoder(resp.Body).Decode(&page)
}

// RebuildPage pulls the latest changes from the page's git repository (for
// git-type pages), re-extracts from the container image (for container_image
// pages), or returns an error for archive pages (which must be rebuilt via
// UploadPageArchive).
//
// Parameters:
//   - name: the page name to rebuild.
//
// Calls POST /pages/rebuild on the Control Plane Service.
func (c *SystemdClient) RebuildPage(ctx context.Context, name string) (_ *account.PageSite, err error) {
	pr, pw := io.Pipe()
	go pipeEncode(pw, PageNameRequest{Name: name})

	resp, err := c.postJSON(ctx, "pages/rebuild", pr)
	if err != nil {
		return nil, fmt.Errorf("%w: RebuildPage: %w", ErrHTTPRequest, err)
	}
	defer func() {
		err = errors.Join(err, resp.Body.Close())
	}()

	if resp.StatusCode != http.StatusOK {
		return nil, readProblemDetail(resp, "POST", "pages/rebuild")
	}

	var page account.PageSite
	return &page, json.NewDecoder(resp.Body).Decode(&page)
}

// UploadPageArchive uploads and extracts an archive into the pages directory
// for the named page. Only valid for pages with source_type "archive". The
// page transitions to "active" status on success or "error" on failure.
//
// Parameters:
//   - name: page name to upload content for (required).
//   - archiveReader: reader providing the archive data (required).
//   - filename: archive filename for format detection (e.g., "site.tar.gz").
//     Supported formats: .tar, .tar.gz/.tgz, .tar.bz2/.tbz2, .tar.xz/.txz.
//
// Calls POST /pages/upload on the Control Plane Service.
func (c *SystemdClient) UploadPageArchive(ctx context.Context, name string, archiveReader io.Reader, filename string) (_ *account.PageSite, err error) {
	pr, pw := io.Pipe()

	writer := multipart.NewWriter(pw)
	go func() {
		if err := writer.WriteField("name", name); err != nil {
			pw.CloseWithError(err)
			return
		}
		part, err := writer.CreateFormFile("archive", filename)
		if err != nil {
			pw.CloseWithError(err)
			return
		}
		if _, err := io.Copy(part, archiveReader); err != nil {
			pw.CloseWithError(err)
			return
		}
		pw.CloseWithError(writer.Close())
	}()

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.route("pages/upload"), pr)
	if err != nil {
		return nil, fmt.Errorf("%w: POST pages/upload: %w", ErrNewRequest, err)
	}
	req.Header.Set("Content-Type", writer.FormDataContentType())
	if c.Token != "" {
		req.Header.Set("Authorization", "Bearer "+c.Token)
	}

	resp, err := c.HTTP.Do(req) //nolint:gosec // G704 -- URL from trusted c.URL
	if err != nil {
		return nil, fmt.Errorf("%w: POST pages/upload: %w", ErrHTTPRequest, err)
	}
	defer func() {
		err = errors.Join(err, resp.Body.Close())
	}()

	if resp.StatusCode != http.StatusOK {
		return nil, readProblemDetail(resp, "POST", "pages/upload")
	}

	var page account.PageSite
	return &page, json.NewDecoder(resp.Body).Decode(&page)
}
